File: src/main/java/org/pulcifer/climate/range/ClimateRanges.java
package org.pulcifer.climate.range;

import org.pulcifer.climate.model.City;
import org.pulcifer.climate.model.ClimateMonth;

import java.math.BigDecimal;
import java.math.MathContext;
import java.math.RoundingMode;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static org.pulcifer.climate.range.ClimateRangeType.AVG_TEMPERATURE;
import static org.pulcifer.climate.range.ClimateRangeType.HIGHEST_TEMPERATURE;
import static org.pulcifer.climate.range.ClimateRangeType.LOWEST_TEMPERATURE;
import static org.pulcifer.climate.range.ClimateRangeType.RAINFALL;
import static org.pulcifer.climate.range.ClimateRangeType.RAIN_DAYS;

public class ClimateRanges {
    private static MathContext context = new MathContext(4, RoundingMode.HALF_UP);
    private static BigDecimal TWO = new BigDecimal("2");
    private Integer numOfOutliersToRemove;
    private Map<ClimateRangeType, Range> ranges = new HashMap<>();

    BigDecimal reduction = new BigDecimal("1");
    BigDecimal inflation = new BigDecimal("1.2");

    BigDecimal rainDayReduction = new BigDecimal("1");
    BigDecimal rainDayInflation = new BigDecimal("1.65");

    BigDecimal rainfallReduction = new BigDecimal("1");
    BigDecimal rainfallInflation = new BigDecimal("1.3");

    int startIdx;
    int endIdx;

    public ClimateRanges (City city, Integer numOfOutliersToRemove) {
        this.numOfOutliersToRemove = numOfOutliersToRemove;
        List<BigDecimal> avgTemps = new ArrayList<>();
        List<BigDecimal> lowTemps = new ArrayList<>();
        List<BigDecimal> highTemps = new ArrayList<>();
        BigDecimal rainfall = BigDecimal.ZERO;
        BigDecimal rainDays = BigDecimal.ZERO;
        for (ClimateMonth m : city.getClimate().getClimateMonth()) {
            avgTemps.add(m.getMinTempF().add(m.getMaxTempF()));
            lowTemps.add(m.getMinTempF());
            highTemps.add(m.getMaxTempF());
            rainfall = rainfall.add(m.getRainfall());
            rainDays = rainDays.add(m.getRaindays());
        }
        startIdx = numOfOutliersToRemove;
        // assuming # of months is 12
        endIdx = 12 - (numOfOutliersToRemove * 2);
        ranges.put(AVG_TEMPERATURE, avgTemperature(avgTemps));
        ranges.put(LOWEST_TEMPERATURE, lowTemperature(lowTemps));
        ranges.put(HIGHEST_TEMPERATURE, highTemperature(highTemps));
        ranges.put(RAIN_DAYS, rainDays(rainDays));
        ranges.put(RAINFALL, rainfall(rainfall));
    }

    private Range avgTemperature(List<BigDecimal> avgTemps) {
        Collections.sort(avgTemps);
        BigDecimal minAvgTemp = avgTemps.get(startIdx).multiply(reduction, context).divide(TWO, context);
        BigDecimal maxAvgTemp = avgTemps.get(endIdx).multiply(inflation, context).divide(TWO, context);
        return new Range(minAvgTemp, maxAvgTemp);
    }

    private Range lowTemperature(List<BigDecimal> lowTemps) {
        Collections.sort(lowTemps);
        BigDecimal minLowestTemp = lowTemps.get(startIdx).multiply(reduction, context);
        BigDecimal maxLowestTemp = lowTemps.get(endIdx).multiply(inflation, context);
        return new Range(minLowestTemp, maxLowestTemp);
    }

    private Range highTemperature(List<BigDecimal> highTemps) {
        Collections.sort(highTemps);
        BigDecimal minHighestTemp = highTemps.get(startIdx).multiply(reduction, context);
        BigDecimal maxHighestTemp = highTemps.get(endIdx).multiply(inflation, context);
        return new Range(minHighestTemp, maxHighestTemp);
    }

    private Range rainDays(BigDecimal rainDays) {
        BigDecimal minRainDays = rainDays.multiply(rainDayReduction, context);
        BigDecimal maxRainDays = rainDays.multiply(rainDayInflation, context);
        return new Range(minRainDays, maxRainDays);
    }

    private Range rainfall(BigDecimal rainfall) {
        BigDecimal minRainfall = rainfall.multiply(rainfallReduction, context);
        BigDecimal maxRainfall = rainfall.multiply(rainfallInflation, context);
        return new Range(minRainfall, maxRainfall);
    }

    public Integer getNumOfOutliersToRemove() { return numOfOutliersToRemove; }
    public Range getAvgTemp() { return ranges.get(AVG_TEMPERATURE); }
    public Range getHighestTemp() { return ranges.get(HIGHEST_TEMPERATURE); }
    public Range getLowestTemp() { return ranges.get(LOWEST_TEMPERATURE); }
    public Range getRainDays() { return ranges.get(RAIN_DAYS); }
    public Range getRainfall() { return ranges.get(RAINFALL); }
}

File: src/main/kotlin/org/pulcifer/climate/range/ClimateRanges.kt
package org.pulcifer.climate.range

import org.pulcifer.climate.model.City
import java.math.BigDecimal
import java.util.Collections
import java.math.MathContext
import java.util.HashMap
import java.math.RoundingMode
import java.util.ArrayList

class ClimateRanges(city: City, val numOfOutliersToRemove: Int? = 0) {
    private val ranges: MutableMap<ClimateRangeType, Range> = HashMap()
    var reduction = BigDecimal(".9")
    var inflation = BigDecimal("1.2")
    var rainDayReduction = BigDecimal(".8")
    var rainDayInflation = BigDecimal("1.65")
    var rainfallReduction = BigDecimal(".8")
    var rainfallInflation = BigDecimal("1.3")
    var startIdx: Int
    var endIdx: Int
    private fun avgTemperature(avgTemps: List<BigDecimal>): Range {
        Collections.sort(avgTemps)
        val minAvgTemp = avgTemps[0].multiply(reduction, context).divide(TWO, context)
        val maxAvgTemp = avgTemps[avgTemps.size - 1].multiply(inflation, context).divide(TWO, context)
        return Range(minAvgTemp, maxAvgTemp)
    }

    private fun lowTemperature(lowTemps: List<BigDecimal?>): Range {
        Collections.sort(lowTemps)
        val minLowestTemp = lowTemps[0]!!.multiply(reduction, context)
        val maxLowestTemp = lowTemps[lowTemps.size - 1]!!.multiply(inflation, context)
        return Range(minLowestTemp, maxLowestTemp)
    }

    private fun highTemperature(highTemps: List<BigDecimal?>): Range {
        Collections.sort(highTemps)
        val minHighestTemp = highTemps[0]!!.multiply(reduction, context)
        val maxHighestTemp = highTemps[highTemps.size - 1]!!.multiply(inflation, context)
        return Range(minHighestTemp, maxHighestTemp)
    }

    private fun rainDays(rainDays: BigDecimal): Range {
        val minRainDays = rainDays.multiply(rainDayReduction, context)
        val maxRainDays = rainDays.multiply(rainDayInflation, context)
        return Range(minRainDays, maxRainDays)
    }

    private fun rainfall(rainfall: BigDecimal): Range {
        val minRainfall = rainfall.multiply(rainfallReduction, context)
        val maxRainfall = rainfall.multiply(rainfallInflation, context)
        return Range(minRainfall, maxRainfall)
    }

    val avgTemp: Range?
        get() = ranges[ClimateRangeType.AVG_TEMPERATURE]
    val highestTemp: Range?
        get() = ranges[ClimateRangeType.HIGHEST_TEMPERATURE]
    val lowestTemp: Range?
        get() = ranges[ClimateRangeType.LOWEST_TEMPERATURE]
    val rainDays: Range?
        get() = ranges[ClimateRangeType.RAIN_DAYS]
    val rainfall: Range?
        get() = ranges[ClimateRangeType.RAINFALL]

    companion object {
        private val context = MathContext(4, RoundingMode.HALF_UP)
        private val TWO = BigDecimal("2")
    }

    init {
        val avgTemps: MutableList<BigDecimal> = ArrayList()
        val lowTemps: MutableList<BigDecimal?> = ArrayList()
        val highTemps: MutableList<BigDecimal?> = ArrayList()
        var rainfall = BigDecimal.ZERO
        var rainDays = BigDecimal.ZERO
        for (m in city!!.climate!!.climateMonth!!) {
            avgTemps.add(m!!.minTempF!!.add(m!!.maxTempF))
            lowTemps.add(m.minTempF)
            highTemps.add(m.maxTempF)
            rainfall = rainfall.add(m.rainfall)
            rainDays = rainDays.add(m.raindays)
        }
        startIdx = numOfOutliersToRemove ?: 0
        // assuming # of months is 12
        endIdx = 12 - (numOfOutliersToRemove ?: 0) * 2
        ranges[ClimateRangeType.AVG_TEMPERATURE] = avgTemperature(avgTemps)
        ranges[ClimateRangeType.LOWEST_TEMPERATURE] = lowTemperature(lowTemps)
        ranges[ClimateRangeType.HIGHEST_TEMPERATURE] = highTemperature(highTemps)
        ranges[ClimateRangeType.RAIN_DAYS] = rainDays(rainDays)
        ranges[ClimateRangeType.RAINFALL] = rainfall(rainfall)
    }
}